### PR TITLE
pd: support pd list.

### PR DESCRIPTION
**Pull request by siddontang**
@ngaut @disksing

### CODE REVIEW

File: src/pd/client.rs [hunk 1]
```diff
@@ -30,31 +30,95 @@ pub trait TRpcClient: Sync + Send {
...
 #[derive(Debug)]
+struct RpcClientCore {
+    addrs: Vec<String>,
+    next_index: usize,
+    stream: Option<TcpStream>,
+}
+
+fn connect(addr: &str) -> Result<TcpStream> {
+    let stream = try!(make_std_tcp_conn(addr));
+    try!(stream.set_read_timeout(Some(Duration::from_secs(SOCKET_READ_TIMEOUT))));
```
Comment (ngaut) [inline src/pd/client.rs]: We should set read timeout or write time out before each read or write.

File: src/pd/client.rs [hunk 2]
```diff
@@ -30,31 +30,95 @@ pub trait TRpcClient: Sync + Send {
...
+            addrs: addrs,
+            next_index: 0,
+            stream: None,
+        }
+    }
+
+    fn try_connect(&mut self) -> Result<()> {
+        let index = self.next_index;
+        self.next_index += 1;
+        if self.next_index >= self.addrs.len() {
```
Comment (ngaut) [inline src/pd/client.rs]: Use self.next_index = self.next_index % self.addrs.len() instead of these three lines.

File: src/pd/client.rs [hunk 3]
```diff
@@ -30,31 +30,95 @@ pub trait TRpcClient: Sync + Send {
 }
 
 #[derive(Debug)]
+struct RpcClientCore {
+    addrs: Vec<String>,
+    next_index: usize,
```
Comment (ngaut) [inline src/pd/client.rs]: Add comment "round-robin"

File: src/pd/client.rs [hunk 4]
```diff
@@ -243,4 +313,80 @@ mod tests {
...
+        let l = TcpListener::bind("127.0.0.1:0").unwrap();
+
+        let addr = format!("{}", l.local_addr().unwrap());
+
+        let h = thread::spawn(move || {
+            loop {
+                let (mut stream, _) = l.accept().unwrap();
+
+                if quit.load(Ordering::Relaxed) {
+                    drop(stream);
```
Comment (disksing) [inline src/pd/client.rs]: Is this `drop(stream)` necessary?

Comment (siddontang) [inline src/pd/client.rs]: maybe not. I will remove it.

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @ngaut

**Comment by siddontang** [comment]
PTAL @BusyJay

**Comment by disksing** [comment]
LGTM

File: src/pd/client.rs [hunk 5]
```diff
@@ -243,4 +313,79 @@ mod tests {
...
+
+                if quit.load(Ordering::Relaxed) {
+                    return;
+                }
+
+                stream.set_nodelay(true).unwrap();
+
+                let leader = leader_index.load(Ordering::Relaxed);
+                if leader != index {
+                    drop(stream);
```
Comment (siddontang) [inline src/pd/client.rs]: @BusyJay  need drop here?

File: src/pd/client.rs [hunk 6]
```diff
@@ -30,31 +30,94 @@ pub trait TRpcClient: Sync + Send {
...
+    try!(rpc::encode_msg(stream, msg_id, message));
+
+    try!(stream.set_read_timeout(Some(Duration::from_secs(SOCKET_READ_TIMEOUT))));
+    let (id, data) = try!(rpc::decode_data(stream));
+    Ok((id, data))
+}
+
+impl RpcClientCore {
+    fn new(dsn: &str) -> RpcClientCore {
+        let addrs: Vec<String> = dsn.split(',').map(String::from).collect();
```
Comment (BusyJay) [inline src/pd/client.rs]: `From::from` seems shorter.

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**